### PR TITLE
fix: harden Pesapal payment processing

### LINKED ISSUE #1593: [Bug]: Harden Pesapal payment verification and duplicate callback handling
### Preflight Checklist

- [x] I have read the contributing guidelines.
- [x] I agree to follow the Code of Conduct.
- [x] I searched the issue tracker and found no open issue covering these Pesapal payment-safety defects.

### Problem Description

The Pesapal API v3 integration can process the same successful payment through multiple entry points: IPN callbacks, browser callbacks, and manual verification. The current success path does not atomically guard dispatch, so overlapping or replayed requests can enqueue payment processing more than once.

A completed status is also accepted without checking that Pesapal's authoritative merchant reference, currency, and amount match the stored transaction.

The IPN response does not return Pesapal's documented acknowledgment fields and status value. In addition, updating IPN registration metadata on an in-memory decrypted credential model can persist dirty plaintext consumer credentials.

### Proposed Solution

- Return the documented IPN acknowledgment for GET and POST notifications.
- Before crediting a completed payment, compare merchant reference, currency, and amount with the stored transaction.
- Lock the Pesapal transaction and persist the success transition before dispatch so replayed or concurrent entry points enqueue processing once.
- Update IPN registration metadata without persisting decrypted credential fields.
- Add focused regression tests for status handling, mismatches, replay behavior, credential encryption, API endpoint selection, and tenant isolation.

### Scope

This change preserves the existing Pesapal API and hosted-checkout flow. It does not add a new payment provider or change production activation behavior.

### PR DESCRIPTION
**Pull request by tushabe**
### Brief summary of the change made

Closes: #1593

This PR hardens the existing Pesapal API v3 payment flow without changing its public API or hosted-checkout behavior.

- Return Pesapal's documented acknowledgment body for GET and POST IPN notifications.
- Require the authoritative completed-payment response to match the stored merchant reference, currency, and amount before crediting.
- Lock the tenant transaction and persist the successful state transition before dispatching `ProcessPayment`, making replayed IPNs, browser callbacks, and verification requests idempotent.
- Prevent decrypted consumer credentials from being written back as plaintext while saving IPN registration metadata.
- Add regression coverage for endpoint selection, token caching, IPN registration and acknowledgment, UGX order creation, status transitions, payment mismatches, callback replay, credential encryption, and tenant isolation.

### Are there any other side effects of this change that we should be aware of?

A completed Pesapal status that does not match the stored merchant reference, currency, or amount is now rejected and logged instead of being credited.

The successful Pesapal status transition and job dispatch execute inside the same tenant database transaction. A dispatch exception rolls the state transition back, while concurrent and replayed requests serialize on the transaction row.

No API routes, database columns, credential response fields, or checkout interfaces are changed.

### Describe how you tested your changes?

- Pesapal PHPUnit suite: 33 tests, 100 assertions, all passing.
- Targeted PHPStan analysis for Pesapal and tenant resolution: no errors.
- PHP CS Fixer: passed.
- Verified GET and POST IPN acknowledgment bodies.
- Verified completed, failed, reversed, invalid, and API-error status handling.
- Verified merchant-reference, currency, and amount mismatches do not dispatch payment processing.
- Verified replaying the successful status path dispatches exactly one `ProcessPayment` job.
- Verified saved consumer credentials remain encrypted at rest and are excluded from API resources.
- Verified a request resolved for one tenant cannot access a Pesapal transaction belonging to another tenant.

### AI assistance disclosure

Codex helped inspect the existing Pesapal implementation, identify payment-safety edge cases, draft the initial code and tests, and prepare this issue and PR description. I reviewed the resulting behavior against Pesapal's API documentation, adapted it to MPM's tenant and payment-processing model, and ran the validation listed above. I remain responsible for the submitted changes and will handle maintainer discussion directly.

### Pull Request checklist

- [x] Meaningful Pull Request title and description
- [x] Changes tested as described above
- [x] Added appropriate test coverage; no user-facing documentation change is required
- [x] Created GitHub issue #1593 for the defects addressed by this PR

### CODE REVIEW

**Comment by tushabe** [comment]
@beesaferoot @AbogeJr ready for review.

File: src/backend/app/Plugins/PesapalPaymentProvider/Services/PesapalTransactionService.php [hunk 1]
```diff
@@ -145,16 +145,46 @@ public function getAll(?int $limit = null): Collection|LengthAwarePaginator {
...
+                throw new \RuntimeException('PesaPal transaction is missing its MPM transaction.');
+            }
+
+            $lockedTransaction->status = PesapalTransaction::STATUS_SUCCESS;
+            $lockedTransaction->save();
+
+            dispatch(new ProcessPayment($companyId, $mpmTransaction->id));
+        });
+
+        $transaction->refresh();
```
Comment (AbogeJr) [inline src/backend/app/Plugins/PesapalPaymentProvider/Services/PesapalTransactionService.php:184]: Nice👍. This fixes a lot of potential issues we would have had. 
Thanks!

**Review by AbogeJr — APPROVED** [review]
@beesaferoot  LGTM! 
Thanks for the improvements and the detailed PR @tushabe

File: src/backend/app/Plugins/PesapalPaymentProvider/Http/Controllers/PesapalPublicController.php [hunk 2]
```diff
@@ -159,7 +159,9 @@ public function showResult(Request $request, string $companyHash, ?int $companyI
                     'created_at' => $transaction->getAttribute('created_at'),
                 ],
                 'verification' => $verification,
-                'success' => isset($verification['status_code']) && $verification['status_code'] === 1,
+                'success' => $verification['error'] === null
```
Comment (beesaferoot) [inline src/backend/app/Plugins/PesapalPaymentProvider/Http/Controllers/PesapalPublicController.php]: Nit: Is it possible to move this to a status variable, `successfulTxCheck` or some similar. I think it makes it easier to read.

File: src/backend/app/Plugins/PesapalPaymentProvider/Services/PesapalTransactionService.php [hunk 3]
```diff
@@ -171,7 +201,26 @@ public function syncStatusFromApi(PesapalTransaction $transaction, int $companyI
...
+            $mismatch = $this->completedPaymentMismatch($transaction, $status);
+            if ($mismatch !== null) {
+                $status['error'] = $mismatch;
+
+                logger()->warning('PesaPal completed payment did not match the stored transaction', [
+                    'pesapal_transaction_id' => $transaction->id,
+                    'order_tracking_id' => $transaction->order_tracking_id,
+                    'mismatch' => $mismatch,
+                ]);
+
```
Comment (beesaferoot) [inline src/backend/app/Plugins/PesapalPaymentProvider/Services/PesapalTransactionService.php]: Can you refactor this to use the logging convention of the codebase i.e `use Illuminate\Support\Facades\Log;`

**Review by beesaferoot — APPROVED** [review]
